ctms/acoustic_service.py

```diff
@@ -0,0 +1,273 @@
+        # populate with all the sub_flags set to false
+        # they'll get set to true below, as-needed
+        acoustic_main_table = {
+            v: "0" for v in AcousticResources.MAIN_TABLE_SUBSCR_FLAGS.values()
+        }
+
+        # flatten the "sub-dicts" into the main dict to match Acoustic's schema
+        for subdict in ["amo", "fxa", "vpn_waitlist"]:
+            for key, value in ctms_data[subdict].items():
+                acoustic_field_name = subdict + "_" + key
+                if key == "fxa_id":
+                    acoustic_field_name = "fxa_id"
+                if (
+                    acoustic_field_name
+                    in AcousticResources.VALID_ACOUSTIC_MAIN_TABLE_FIELDS
+                ):
+                    acoustic_main_table[
+                        acoustic_field_name
+                    ] = self._transform_field_for_acoustic(value)
+                else:
+                    print(
+                        "Skipping CTMS field ctms_data[%s][%s] because no match in Acoustic"
+                        % (subdict, key)
+                    )
+
+        for key, value in ctms_data["email"].items():
+            acoustic_field_name = key
+            if acoustic_field_name == "primary_email":
+                acoustic_field_name = "email"
+            if (
+                acoustic_field_name
+                in AcousticResources.VALID_ACOUSTIC_MAIN_TABLE_FIELDS
+            ):
+                acoustic_main_table[
+                    acoustic_field_name
+                ] = self._transform_field_for_acoustic(value)
+            else:
+                print(
+                    "Skipping CTMS field ctms_data[email][%s] because no match in Acoustic"
+                    % (key)
+                )
+
+        # create the RT rows for the newsletter table in acoustic
+        newsletter_rows = []
+        for newsletter in ctms_data["newsletters"]:
+            newsletter_rows.append(
+                {
+                    "email_id": ctms_data["email"]["email_id"],
+                    "newsletter_name": newsletter["name"],
+                    "newsletter_format": newsletter["format"],
+                    "newsletter_lang": newsletter["lang"],
+                    "newsletter_source": newsletter["source"],
+                    "newsletter_unsub_reason": newsletter["unsub_reason"],
+                    "create_date": self._transform_field_for_acoustic(
+                        newsletter["create_timestamp"]
+                    ),
+                    "update_date": self._transform_field_for_acoustic(
+                        newsletter["update_timestamp"]
+                    ),
+                }
+            )
+            # and finally flip the main table's sub_<newsletter> flag to true for each subscription
+            if newsletter["subscribed"]:
+                acoustic_main_table[
+                    AcousticResources.MAIN_TABLE_SUBSCR_FLAGS[newsletter["name"]]
+                ] = "1"
```

Comment:
this won't necessarily always be true in ctms I think, right? This was true in an initial sync out of bq but not in the general case

Reply:
I don't understand this comment. _What_ won't necessarily be true in ctms?

Reply:
"flip the main table's sub_<newsletter> flag to true for each subscription"

won't we have "subscribed: False" in our tables?

Reply:
Yes, and those won't get flipped. At least, assuming that field as given to us as a boolean. If it's just a string, then that'd be trouble.